File: kotlin/src/main/kotlin/datastructures/Queue.kt
package datastructures

class Queue<T> {
    private var head : QNode<T>? = null
    private var tail: QNode<T>? = null
    private var size: Int = 0

    fun size(): Int = size

    fun add(value: T){
        var node = QNode(value)
        var currentTail = tail

        if (currentTail != null) {
            currentTail.next = node;
        }

        if (head == null) {
            this.head = node;
        }
        tail = node
        size++
    }

    fun remove(): T? {
        /*var currentHead = head

        if (currentHead == null) {
            return null
        }*/
        var currentHead: QNode<T>? = head ?: return null

        head = currentHead?.next

        size--
        return currentHead?.data
    }

}

data class QNode<T>(var data: T, var next: QNode<T>? = null)

File: go/datastructures/queue.go
package datastructures

type Queue struct {
	head *Node
	tail *Node
	size int
}

func (queue Queue) Size() int {
	return queue.size
}

func (queue Queue) add(value interface{}) {
	node := &Node{
		Data: value,
		Next: nil,
	}

	tail := queue.tail

	if tail != nil {
		tail.Next = node
	}

	if queue.head == nil {
		queue.head = node
	}

	queue.tail = node
	queue.size++
}
func (queue Queue) remove() interface{} {
	head := queue.head

	if head == nil {
		return nil
	}

	queue.head = head.Next

	queue.size--

	return head.Data
}
